homework/solutions/nelson/session1-spider1.py

```diff
@@ -0,0 +1,102 @@
+from asyncio.futures import Future
+from asyncio.tasks import sleep
+import logging
+from typing import List
+import asyncio
+import sys
+
+import aiohttp
+from typing_extensions import IntVar
+
+logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+async def get(url: str = None, counter: int = 0, callback = None, session: aiohttp.ClientSession = None) -> str:
+    """
+    Get a webpage's content, given it's URL. Uses only a simple GET HTTP Request
+
+    :param str url: A fully qualified URL to fetch.
+    :param int counter: Which call is this? Order matters!
+    :param aiohttp.ClientSession session: A re-usable session for making calls.
+
+    :return: str The result.
+    """
+    if session is None:
+        raise ValueError('`session` is a required argument. You must pass an HTTP Session')
+    logger.info(f'Awaiting {url} in call {counter}.')
+    result = await session.get(url)
+    if callback:
+        callback(url, counter, await result.text())
+    return await result.text()
+
+def get_using_task(url: str, counter: int, callback = None, session : aiohttp.ClientSession = None) -> asyncio.Task:
+    if session is None:
+        raise ValueError('`session` is a required argument. You must pass an HTTP Session')
+
+    mycoroutine = get(url, counter, callback, session)
+    return asyncio.create_task(mycoroutine)
+
+async def spider(urls: List[str], concurrency: int = 1, callback = None, session : aiohttp.ClientSession = None):
+    if session is None:
+        raise ValueError('`session` is a required argument. You must pass an HTTP Session')
+
+    complete_tasks = set()
+    pending_tasks = set()
+
+    for counter, url in enumerate(urls):
+        logger.info(f'Spider[{counter+1}]: {url}')
+
+        # adds new task per each iteration
+        pending_tasks.add(get_using_task(url, counter=counter+1, callback=callback, session=session))
+        logger.info(f'Adding tasks => pending_tasks: {len(pending_tasks)}')
+
+        # pendings_tasks already reached the max value (concurrency), executes the tasks and waits for the next one to complete
+        if len(pending_tasks) >= concurrency:
+            logger.info('Executing tasks...')
+            done, pending_tasks = await asyncio.wait(pending_tasks, return_when=asyncio.FIRST_COMPLETED)
+            complete_tasks.update(done)
+            logger.info(f'Executing tasks => completed_tasks={len(done)} pending_tasks={len(pending_tasks)}')
+
+        # Last iteration: complete ALL the remaining pending_tasks (if existent)
+        if counter == len(urls)-1 and len(pending_tasks) > 0:
```

Comment:
Would pulling this final stage out of the loop allow you to get rid of the conditional statement?